File: packages/java/anip-service/src/main/java/dev/anip/service/StreamResult.java
package dev.anip.service;

import java.util.concurrent.BlockingQueue;

/**
 * Holds the result of a streaming invocation.
 * The events queue emits progress events followed by exactly one terminal event
 * (completed or failed), then a poison pill (type="__done__") to signal completion.
 * Call cancel to signal the handler that the client has disconnected.
 */
public class StreamResult {

    /** Poison pill event type to signal end of stream. */
    public static final String DONE_TYPE = "__done__";

    private final BlockingQueue<StreamEvent> events;
    private final Runnable cancel;

    public StreamResult(BlockingQueue<StreamEvent> events, Runnable cancel) {
        this.events = events;
        this.cancel = cancel;
    }

    public BlockingQueue<StreamEvent> getEvents() {
        return events;
    }

    public Runnable getCancel() {
        return cancel;
    }
}

File: packages/csharp/src/Anip.Service/StreamResult.cs
using System.Threading.Channels;

namespace Anip.Service;

/// <summary>
/// Holds a channel of streaming events.
/// The channel is closed after exactly one terminal event (completed or failed).
/// Call Cancel() to signal the handler that the client has disconnected.
/// </summary>
public class StreamResult
{
    public ChannelReader<StreamEvent> Events { get; }
    public Action Cancel { get; }

    public StreamResult(ChannelReader<StreamEvent> events, Action cancel)
    {
        Events = events;
        Cancel = cancel;
    }
}
